core/storage/src/storages/mod.rs

```diff
@@ -307,35 +307,39 @@ fn extend_with_cache<T: Clone + Encode>(
     storage.write(db_transaction)?;
     // If it has reached here then it is safe to put in cache.
     for (key, value) in cache_to_extend {
-        cache.put(key, value);
+        cache.cache_set(key, value);
     }
     Ok(())
 }
 
 fn read_with_cache<'a, T: Decode + Clone + 'a>(
     storage: &KeyValueDB,
     col: u32,
-    cache: &'a mut LruCache<Vec<u8>, T>,
+    cache: &'a mut SizedCache<Vec<u8>, T>,
     key: &[u8],
 ) -> StorageResult<&'a T> {
     let key_vec = key.to_vec();
-    if cache.contains(&key_vec) {
-        let value = cache.get(&key_vec).unwrap();
-        Ok(Some(value))
-    } else if let Some(data) = storage.get(Some(col), key)? {
-        let result = Decode::decode(data.as_ref())?;
-        cache.put(key_vec.clone(), result);
-        Ok(cache.get(&key_vec))
-    } else {
-        Ok(None)
+    // This code block is safe. Has to use unsafe due to Cached API limitations,
+    // mainly that there is not lookup method that takes &self.
+    unsafe {
```

Comment:
This should be resolvable without unsafe. Here is the minimum reproducible example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b0bebdb3467a12d99b1b219dce688b05 I asked in Rust discord why Rust NLL does not resolve this.

Reply:
Here is my suboptimal solution with double lookup. I suggest we go with it instead, unless we later determine that it creates a performance bottleneck.
```
    let key_vec = key.to_vec();
    if cache.cache_get(&key_vec).is_some() {
        return Ok(Some(cache.cache_get(&key_vec).unwrap()));
    }
    if let Some(data) = storage.get(Some(col), key)? {
        let result = Decode::decode(data.as_ref())?;

        let key_vec = key_vec.to_vec();
        cache.cache_set(key_vec.to_vec(), result);
        return Ok(cache.cache_get(&key_vec));
    }
    Ok(None)
}
```

Reply:
Rust gurus answered -- it is the current limitation of compiler. I suggest we go with suboptimal oppose to unsafe.

Reply:
I don't think not using unsafe is possible given that `cache_get` takes `&mut self` instead of `&self`.

Reply:
The above code that I pasted is working.

Reply:
I see. I misread.